Add the scripting property to the browser namespace

kiaraarose opened the pull request:
#### dd38ff3650973c7f09ba7fd1a8cf6a2abf131b58
<pre>
Add the scripting property to the browser namespace
<a href="https://bugs.webkit.org/show_bug.cgi?id=259948">https://bugs.webkit.org/show_bug.cgi?id=259948</a>

Reviewed by Timothy Hatcher.

This patch adds initial support for `browser.scripting` by adding the APIs to the idl file and
doing data validation in the Web Process.

* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources-output.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPINamespaceCocoa.mm:
(WebKit::WebExtensionAPINamespace::scripting):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIScriptingCocoa.mm: Added.
(WebKit::WebExtensionAPIScripting::executeScript):
(WebKit::WebExtensionAPIScripting::insertCSS):
(WebKit::WebExtensionAPIScripting::removeCSS):
(WebKit::WebExtensionAPIScripting::registerContentScripts):
(WebKit::WebExtensionAPIScripting::getRegisteredContentScripts):
(WebKit::WebExtensionAPIScripting::updateContentScripts):
(WebKit::WebExtensionAPIScripting::unregisterContentScripts):
(WebKit::WebExtensionAPIScripting::validateScript):
(WebKit::WebExtensionAPIScripting::validateTarget):
(WebKit::WebExtensionAPIScripting::validateCSS):
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPINamespace.h:
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIPermissions.h:
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIScripting.h:
Copied from Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIPermissions.h.
* Source/WebKit/WebProcess/Extensions/Cocoa/WebExtensionContextProxyCocoa.mm:
* Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPINamespace.idl:
* Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPIScripting.idl:
Copied from Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPINamespace.idl.
* Tools/TestWebKitAPI/SourcesCocoa.txt:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIScripting.mm: Added.
(TestWebKitAPI::TEST):

Canonical link: <a href="https://commits.webkit.org/268255@main">https://commits.webkit.org/268255@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/034c1bd52f57429eca70caf53fcc4b2f15ee2065

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/19163 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/19575 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/20172 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/21052 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/17937 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/19378 "Passed tests") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/23/builds/22850 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/19703 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/21052 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/19384 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/23/builds/22850 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/16663 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/21929 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/23/builds/22850 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/17450 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/21929 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/23/builds/22850 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/17624 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/21929 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/7/builds/18220 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/15433 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/30/builds/17340 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/19/builds/4581 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/4/builds/21701 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/2333 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/31/builds/18056 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

kiaraarose commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIScriptingCocoa.mm:
```diff
@@ -0,0 +1,254 @@
+            *outExceptionString = toErrorString(nil, frameIDsKey, @"at least one frame id is not a positive number.");
+            return false;
+        }
+    }
```
newline below

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIScriptingCocoa.mm:
```diff
@@ -0,0 +1,255 @@
+#import "config.h"
+#import "WebExtensionAPIScripting.h"
+
+#import "CocoaHelpers.h"
+#import "Logging.h"
+#import "MessageSenderInlines.h"
+#import "WebExtension.h"
+#import "WebExtensionAPINamespace.h"
+#import "WebExtensionContextMessages.h"
+#import "WebExtensionUtilities.h"
+#import "WebProcess.h"
```
These should all move down inside the `#if ENABLE(WK_WEB_EXTENSIONS)`

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIScriptingCocoa.mm:
```diff
@@ -0,0 +1,255 @@
+        if (frameID.longLongValue < mainFrameID) {
+            *outExceptionString = toErrorString(nil, frameIDsKey, @"at least one frame id is not a positive number.");
+            return false;
+        }
```
I'm adding a helper for parsing frameIDs soon, it checks more than negative. You will want to use that here once it is added.

kiaraarose replied:
will add a fixme

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIScriptingCocoa.mm:
```diff
@@ -0,0 +1,255 @@
+        return false;
+
+    if (targetInfo[allFramesKey] && targetInfo[frameIDsKey]) {
+        *outExceptionString = toErrorString(nil, nil, @"The 'allFrames' and 'frameIds' properties cannot both be specified.");
```
```
*outExceptionString = toErrorString(nil, targetKey, @"it cannot specify both 'allFrames' and 'frameIds'");
```

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIScriptingCocoa.mm:
```diff
@@ -0,0 +1,255 @@
+    NSArray *frameIDs = targetInfo[frameIDsKey];
+    for (NSNumber *frameID in frameIDs) {
+        if (frameID.longLongValue < mainFrameID) {
+            *outExceptionString = toErrorString(nil, frameIDsKey, @"at least one frame id is not a positive number.");
```
Can drop the trailing period, the helper function adds it if needed.

xeenon replied:
I would just say `@"it is not a frame identifier"` to cover all the bases. This matches tab and window ID errors.

xeenon replied:
Better yet, since this is an array, I have been including the problem value.

```
toErrorString(nil, frameIDsKey, @"'%@' is not a frame identifier", frameID)
```

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIScriptingCocoa.mm:
```diff
@@ -0,0 +1,255 @@
+
+    if (NSString *world = script[worldKey]) {
+        if (![world isEqualToString:isolatedWorld] && ![world isEqualToString:mainWorld]) {
+            *outExceptionString = toErrorString(nil, worldKey, @"it must either have the value of 'ISOLATED' or 'MAIN'.");
```
```
*outExceptionString = toErrorString(nil, worldKey, @"it must specify either 'ISOLATED' or 'MAIN'");
```

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIScriptingCocoa.mm:
```diff
@@ -0,0 +1,255 @@
+    if (script[filesKey]) {
+        NSArray *files = script[filesKey];
+        if (!files.count) {
+            *outExceptionString = toErrorString(nil, filesKey, @"at least one file must be passed.");
```
Drop the trailing period.

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIScriptingCocoa.mm:
```diff
@@ -0,0 +1,255 @@
+        return false;
+    }
+
+    BOOL scriptContainsArguments = !!(script[argsKey] || script[argumentsKey]);
```
`bool`

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIScriptingCocoa.mm:
```diff
@@ -0,0 +1,255 @@
+
+    BOOL scriptContainsArguments = !!(script[argsKey] || script[argumentsKey]);
+    if (scriptContainsArguments && !functionWasPassed) {
+        *outExceptionString = toErrorString(nil, nil, @"The 'args' property must be passed along with the 'func' property.");
```
```
*outExceptionString = toErrorString(nil, @"details", @"it must specify both 'func' and 'args'");
```

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIScriptingCocoa.mm:
```diff
@@ -0,0 +1,255 @@
+
+    bool functionWasPassed = script[functionKey] || script[funcKey];
+    if ((script[filesKey] && functionWasPassed)
+        || (!functionWasPassed && !script[filesKey])) {
```
One line would be easier to read.

kiaraarose replied:
oops, didn't mean to put on two lines, thanks!

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIScriptingCocoa.mm:
```diff
@@ -0,0 +1,255 @@
+    static NSArray<NSString *> *requiredKeys = @[
+        tabIDKey,
+    ];
+    static NSDictionary<NSString *, id> *keyTypes = @{
```
Newline above.

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIScriptingCocoa.mm:
```diff
@@ -0,0 +1,255 @@
+    }
+
+    if (script[argumentsKey] && script[argsKey]) {
+        *outExceptionString = toErrorString(nil, nil, @"Both 'args' and 'arguments' cannot be passed. Please use 'args'.");
```
```
*outExceptionString = toErrorString(nil, @"details", @"it cannot specify both 'args' and 'arguments'. Please use 'args'")
```

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIScriptingCocoa.mm:
```diff
@@ -0,0 +1,255 @@
+        return false;
+
+    if (script[functionKey] && script[funcKey]) {
+        *outExceptionString = toErrorString(nil, nil, @"Both 'func' and 'function' cannot be passed. Please use 'func'.");
```
```
*outExceptionString = toErrorString(nil, @"details", @"it cannot specify both 'func' and 'function'. Please use 'func'")
```

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIScriptingCocoa.mm:
```diff
@@ -0,0 +1,255 @@
+        worldKey: NSString.class,
+    };
+
+    if (!validateDictionary(script, nil, requiredKeys, keyTypes, outExceptionString))
```
The parameter name should be used for this, which is `details`, IIRC.

`validateDictionary(script, @"details", requiredKeys, keyTypes, outExceptionString)`

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIScriptingCocoa.mm:
```diff
@@ -0,0 +1,255 @@
+    bool functionWasPassed = script[functionKey] || script[funcKey];
+    if ((script[filesKey] && functionWasPassed)
+        || (!functionWasPassed && !script[filesKey])) {
+        *outExceptionString = toErrorString(nil, nil, @"Either 'func' or 'files' must be passed, but not both at the same time.");
```
```
*outExceptionString = toErrorString(nil, @"details", @"it cannot specify both 'func' and 'files'")
```

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIScriptingCocoa.mm:
```diff
@@ -0,0 +1,255 @@
+
+    if ((cssInfo[cssKey] && cssInfo[filesKey])
+        || (!cssInfo[filesKey] && !cssInfo[cssKey])) {
+        *outExceptionString = toErrorString(nil, nil, @"Either 'css' or 'files' must be passed, but not both at the same time.");
```
```
*outExceptionString = toErrorString(nil, targetKey, @"it cannot specify both 'css' and 'files'");
```

kiaraarose replied:
will split this into two checks so that the error message can be `it cannot specify both 'css' and 'files'` or `either 'css' or 'files' must be passed`

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIScriptingCocoa.mm:
```diff
@@ -0,0 +1,255 @@
+        return false;
+
+    if ((cssInfo[cssKey] && cssInfo[filesKey])
+        || (!cssInfo[filesKey] && !cssInfo[cssKey])) {
```
One line.

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIScriptingCocoa.mm:
```diff
@@ -0,0 +1,255 @@
+        targetKey: NSDictionary.class,
+    };
+
+    if (!validateDictionary(cssInfo, nil, requiredKeys, keyTypes, outExceptionString))
```
`validateDictionary(cssInfo, @"details", requiredKeys, keyTypes, outExceptionString)`

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIScriptingCocoa.mm:
```diff
@@ -0,0 +1,255 @@
+        allFramesKey: @YES.class,
+    };
+
+    if (!validateDictionary(targetInfo, nil, requiredKeys, keyTypes, outExceptionString))
```
`validateDictionary(targetInfo, targetKey, requiredKeys, keyTypes, outExceptionString)`

kiaraarose commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIScriptingCocoa.mm:
```diff
@@ -0,0 +1,265 @@
+        return false;
+    }
+
+    if (script[filesKey]) {
```
`if (NSArray *files = script[filesKey])`

kiaraarose commented on Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIScripting.h:
```diff
@@ -0,0 +1,64 @@
+
+class WebExtensionAPIScripting : public WebExtensionAPIObject, public JSWebExtensionWrappable {
+    WEB_EXTENSION_DECLARE_JS_WRAPPER_CLASS(WebExtensionAPIScripting, scripting);
+public:
```
newline above

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIScriptingCocoa.mm:
```diff
@@ -0,0 +1,264 @@
+
+    // A JSValue cannot be transferred across processes, so we must convert it to a string before passing it along.
+    if (JSValue *func = script[funcKey] ?: script[functionKey]) {
+        ASSERT(func.isObject);
```
This should be more than an `ASSERT`, since validate only checks NSObject.class currently. This should be an exception if `!func.isObject`.

kiaraarose replied:
Spoke to @xeenon about this offline. Doesn't look like there is a way for anything to pass the `JSValue.class` check but fail `isObject`. Will change the validation to check against `JSValue` instead of `NSObject` but will remove this `isObject` check.

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIScriptingCocoa.mm:
```diff
@@ -0,0 +1,264 @@
+        return;
+
+    // A JSValue cannot be transferred across processes, so we must convert it to a string before passing it along.
+    if (JSValue *func = script[funcKey] ?: script[functionKey]) {
```
This can break currently since `validateDictionary()` only checks `NSObject.class`.

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIScriptingCocoa.mm:
```diff
@@ -0,0 +1,264 @@
+        argumentsKey: @[ NSObject.class ],
+        filesKey: @[ NSString.class ],
+        funcKey: NSObject.class,
+        functionKey : NSObject.class,
```
This will never fail, since everything is an `NSObject`, and `nil` values are excluded from the dictionary. These should really be `JSValue.class` at least, and `classToClassString()` needs modified to have `JSValue.class` strings. Sadly we can't get more specific with a "function" class, so the strings should be "value" or "object". A check later for `func.isObject` can maybe at least say "function".

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIScriptingCocoa.mm:
```diff
@@ -0,0 +1,264 @@
+
+    if (NSString *world = script[worldKey]) {
+        if (![world isEqualToString:isolatedWorld] && ![world isEqualToString:mainWorld]) {
+            *outExceptionString = toErrorString(nil, worldKey, @"it must either have the value of 'ISOLATED' or 'MAIN'");
```
`@"it must specify 'ISOLATED' or 'MAIN'"`

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIScriptingCocoa.mm:
```diff
@@ -0,0 +1,264 @@
+    }
+
+    if (!cssInfo[filesKey] && !cssInfo[cssKey]) {
+        *outExceptionString = toErrorString(nil, @"details", @"either 'css' or 'files' must be specified");
```
`@"it must specify 'css' or 'files'`

kiaraarose commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIScriptingCocoa.mm:
```diff
@@ -0,0 +1,264 @@
+    static NSArray<NSString *> *requiredKeys = @[
+        targetKey,
+    ];
+    static NSDictionary<NSString *, id> *keyTypes = @{
```
newline above

xeenon commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIScripting.mm:
```diff
@@ -0,0 +1,109 @@
+        @"browser.test.assertThrows(() => browser.scripting.removeCSS({target: { tabId: '0' }, files: ['path/to/file'], css: 'css'}), /'tabId' is expected to be a number, but a string was provided./i)",
+
+
+        @"browser.test.assertSafeResolve(() => browser.scripting.executeScript({target: { tabId: 0, allFrames: false }, files: ['path/to/file'], world: 'ISOLATED' }))",
+        @"browser.test.assertSafeResolve(() => browser.scripting.executeScript({target: { tabId: 0 }, func: () => 'function' }))",
+
+        @"browser.test.assertSafeResolve(() => browser.scripting.insertCSS({target: { tabId: 0, allFrames: false }, files: ['path/to/file'] }))",
+        @"browser.test.assertSafeResolve(() => browser.scripting.insertCSS({target: { tabId: 0, frameIds: [0] }, files: ['path/to/file'] }))",
+        @"browser.test.assertSafeResolve(() => browser.scripting.insertCSS({target: { tabId: 0, allFrames: true }, css: 'body { background-color: pink }' }))",
+
+        @"browser.test.assertSafeResolve(() => browser.scripting.removeCSS({target: { tabId: 0, allFrames: false }, files: ['path/to/file'] }))",
+        @"browser.test.assertSafeResolve(() => browser.scripting.removeCSS({target: { tabId: 0, frameIds: [0] }, css: 'body { background-color: pink }' }))",
+        @"browser.test.assertSafeResolve(() => browser.scripting.removeCSS({target: { tabId: 0, allFrames: true }, css: 'body { background-color: pink }' }))",
```
These are good for now, but will fail once you implement since tab not be found. You can remove them when that is the case though!

kiaraarose committed:
Add the scripting property to the browser namespace
https://bugs.webkit.org/show_bug.cgi?id=259948

Reviewed by Timothy Hatcher.

This patch adds initial support for `browser.scripting` by adding the APIs to the idl file and
doing data validation in the Web Process.

* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources-output.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPINamespaceCocoa.mm:
(WebKit::WebExtensionAPINamespace::scripting):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIScriptingCocoa.mm: Added.
(WebKit::WebExtensionAPIScripting::executeScript):
(WebKit::WebExtensionAPIScripting::insertCSS):
(WebKit::WebExtensionAPIScripting::removeCSS):
(WebKit::WebExtensionAPIScripting::registerContentScripts):
(WebKit::WebExtensionAPIScripting::getRegisteredContentScripts):
(WebKit::WebExtensionAPIScripting::updateContentScripts):
(WebKit::WebExtensionAPIScripting::unregisterContentScripts):
(WebKit::WebExtensionAPIScripting::validateScript):
(WebKit::WebExtensionAPIScripting::validateTarget):
(WebKit::WebExtensionAPIScripting::validateCSS):
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPINamespace.h:
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIPermissions.h:
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIScripting.h:
Copied from Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIPermissions.h.
* Source/WebKit/WebProcess/Extensions/Cocoa/WebExtensionContextProxyCocoa.mm:
* Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPINamespace.idl:
* Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPIScripting.idl:
Copied from Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPINamespace.idl.
* Tools/TestWebKitAPI/SourcesCocoa.txt:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIScripting.mm: Added.
(TestWebKitAPI::TEST):

Canonical link: https://commits.webkit.org/268255@main
